Fix: redirect to home page after a paste is deleted

Parthiv-M opened the pull request:
This PR fixes #1156 

## Changes
<!-- List all the changes you have done -->
* Once a paste is deleted, the alert is displayed and the user is redirected to the home page after 5 seconds
* The alert has been changed to be of type `alert-success` (green coloured) if `$STATUS` equals the `Paste was properly deleted.` message; `alert-info` otherwise

Parthiv-M commented:
Hi @rugk, would be great if you can take a look at this patch :)

elrido commented on tpl/bootstrap.php:
```diff
@@ -490,7 +490,7 @@
 endif;
 ?>
-				<div id="status" role="alert" class="alert alert-info<?php echo empty($STATUS) ? ' hidden' : '' ?>">
+				<div id="status" role="alert" class="alert alert-<?php echo $STATUS === 'Paste was properly deleted.' ? 'success' : 'info' ?><?php echo empty($STATUS) ? ' hidden' : '' ?>">
```
I wouldn't change the `#status` box design for this case, as it will then stay in `alert-success` (green) mode even when other informational (blue) messages get displayed.

Parthiv-M replied:
I believe that the check `$STATUS === 'Paste was properly deleted.' ? 'success' : 'info'` will ensure that the alert is green only when it is the paste message being displayed? Otherwise it will still be the original blue alert.

elrido replied:
Granted, right, it will work because you force a reload with the timeout.

elrido commented on tpl/bootstrap5.php:
```diff
@@ -353,7 +353,7 @@
 endif;
 ?>
-				<div id="status" role="alert" class="alert alert-info<?php echo empty($STATUS) ? ' hidden' : '' ?>">
+				<div id="status" role="alert" class="alert alert-<?php $STATUS == 'Paste was properly deleted.' ? 'success' : 'info' ?><?php echo empty($STATUS) ? ' hidden' : '' ?>">
```
Same reasoning as above: I wouldn't change the `#status` box design for this case, as it will then stay in `alert-success` (green) mode even when other informational (blue) messages get displayed.

Parthiv-M replied:
Same justification as above.

elrido reviewed (changes requested):
Apart from not changing the information box colour, I wonder if it would be better UX if the user instead gets to acknowledge the redirect (they currently do that by clicking on "new"). An automatic redirect might lead to them missing the confirmation message if they are distracted during those 5 seconds.

elrido commented:
With these two comments resolved, my only concern remaining is the UX one - forcing a reload without user interaction might hide the message before they see it. For this reason I prefer the current behaviour where the user has to explicitly click on the "New" button to move on to create a new paste (if that is what they want - maybe they rather take a screenshot as proof, or just close the browser tab).

Let's give the other maintainers a chance to speak up on this. If no one is opposed though, I'm willing to merge it, after updating the SRI-hashes and documenting the change in the changelog.

Parthiv-M commented:
I have a suggestion from the UX perspective that could probably be better than the current method of clicking on the PrivateBin logo to go back to the home page: inside the green alert itself, we place a button that says something like `<-- Back home` or `<-- Create new paste` with the `<--` being replaced by a icon.

Happy to hear feedback from the maintainers about this :)

elrido commented:
> [...] inside the green alert itself, we place a button that says something like `<-- Back home` or `<-- Create new paste` with the `<--` being replaced by a icon.

That would be a lot better, IMHO. It could be a button with icon, maybe styled just like the new button? For example, in the bootstrap5.php template:

```php
				<div id="status" role="alert" class="alert alert-<?php echo ($STATUS == 'Paste was properly deleted.' ? 'success' : 'info'), (empty($STATUS) ? ' hidden' : '') ?>">
					<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#info-circle" /></svg>
					<?php echo I18n::encode($STATUS), PHP_EOL; ?>
				</div>
<?php
if ($STATUS == 'Paste was properly deleted.') :
?>
				<button type="button" class="btn btn-secondary flex-fill" onclick="$.PrivateBin.UiHelper.reloadHome()">
					<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#file-earmark" /></svg> <?php echo I18n::_('New'), PHP_EOL; ?>
				</button>
<?php
endif;
?>
```
Aside: I realize now that this will only work for English, as the $STATUS contains an already translated message. We may have to detect this around the below line in the controller and assign some other, boolean flag, say `isDeleted` to the view that we can more easily check in the above logic:
https://github.com/PrivateBin/PrivateBin/blob/d69d29f3a940b443760f1e227df6628e3f4cfb3f/lib/Controller.php#L414

Parthiv-M commented:
Ah, makes sense! Let me try to make these changed and update this patch

Parthiv-M commented:
So, I've made the following changes:
- [x] Included a new status message in the Controller that can track the deletion and pass the deleted status to the template
- [x] Added an event listener for the _New_ button inside the alert, because adding an `onclick` in the template would cause issues with loading the jquery function (due to CSP config)

## Previews

The button **with Bootstrap 5**
![alert-bootstrap5](https://github.com/user-attachments/assets/9a4c0f49-7544-44d4-b369-a172e7e9a879)


The button **with Bootstrap 3** has  a few options, I've pushed the one that looked best to me, but it is up for suggestions/changes :)
- `btn-default` (pushed)
![alert-btn-default](https://github.com/user-attachments/assets/53715c6b-eb6f-4070-b12a-b183eeb3b48e)

- `btn-primary`
![alert-btn-primary](https://github.com/user-attachments/assets/fc5b293c-c9bc-4f97-9679-134a8316be0e)

- `btn-secondary`
![alert-btn-secondary](https://github.com/user-attachments/assets/dab1ae1d-b9be-4ee4-81cf-b4c762ca42bf)

rugk commented:
Did not you want to name it differently than "New"? "Back"? Without context, at this position, it is quite unclear IHMO, what "new" thing can be created here?

Ideas:
* either place it back into the toolbar, where it usually is shown aka the "New" button. People likely will be accustomed to that behavior.
* Or maybe keep it there, but clarify the name more, like "Start over" or maybe just "Okay", which redirects to the home page. Just some kind of general acknowledgement that does not arise questions what new thing could be done or whether the paste will be resurrected… whatever :wink:

rugk commented on tpl/bootstrap.php:
```diff
@@ -489,10 +489,17 @@
-?>
-				<div id="status" role="alert" class="alert alert-info<?php echo empty($STATUS) ? ' hidden' : '' ?>">
+?>					
+				<div id="status" role="alert" class="clearfix alert alert-<?php echo (bool)$ISDELETED ? 'success' : 'info'; echo empty($STATUS) ? ' hidden' : '' ?>">
```
I am curious… What's `clearfix`?

Parthiv-M replied:
It is used to clear floats for the element ([ref](https://getbootstrap.com/docs/3.3/css/#helper-classes-clearfix)) so that the `pull-right` in the child button can work properly.

This is how it looks without `clearfix`
![alert-without-clearfix](https://github.com/user-attachments/assets/c0c334d5-7fbc-473c-8ed4-8357e909e0ae)

ppfeufer replied:
Hint:
The `pull-right` class does no longer exist since Bootstrap 4 » https://stackoverflow.com/questions/39655110/what-happened-to-the-pull-left-and-pull-right-classes-in-bootstrap-4

And since the project moved to Bootstrap 5, this might need an overhaul.

Parthiv-M replied:
That's right. I'm only using the `pull-right` in the Bootstrap 3 template. The Bootstrap 5 doesn't have any of either `clearfix` or `pull-right`

rugk reviewed (commented):
Apart from that, I like the idea. Just needs some testing that it really works in all templates, you know…

Parthiv-M commented:
> Did not you want to name it differently than "New"? 

I want to go for _New Paste_, I'll make that change

Parthiv-M commented:
Additionally, is there something I need to do to update the SRIs? And the changelog? If you can give me links to some docs, I can do that too.

elrido reviewed (changes requested):
I would really appreciate if you could add the changelog entry, it saves me having to find and remember adding it during the release - that is a pretty tedious manual step for me. Other than that, we now only need the SRI hashes updated - running phpunit will do that for you. We do offer a [unit testing container image](https://github.com/PrivateBin/docker-unit-testing/), if you don't have that tool installed on your dev machine.

Parthiv-M commented:
Hi @rugk, I've made the following changes
- Updated the text (to _Start over_) and the icon (to a repeat icon)
- Updated the `CHANGELOG.md` file


I had one query (which is also why one test is failing in the CI): do I have to add keys and values for all the languages (every `<lang-code>.json` file)? If so, I can do that (and push the SRI fix) and everything should be done.

elrido commented on i18n/en.json:
```diff
@@ -26,6 +26,7 @@
     "JavaScript is required for %s to work. Sorry for the inconvenience.": "JavaScript is required for %s to work. Sorry for the inconvenience.",
     "%s requires a modern browser to work.": "%s requires a modern browser to work.",
     "New": "New",
+    "Start over": "Start over",
```
Two things:
- not required, but we did usually add new translations to the end of the file, so the untranslated new items are easy to find for translators
- you would need to add the new message (with English translation) to all files for the tests to work - this prevents exceptions at runtime due to missing translations and prevents the files getting out of sync. I usually use the jq tool to append new items to JSON files and ensure they are all formatted the same.

elrido reviewed (changes requested):
Looks good and thanks for the changelog entry. Once the new translation is in all files it will pass the tests (except for php 8.4 where we wait on a fix in a dependency)

elrido reviewed (dismissed):
Thank you!
